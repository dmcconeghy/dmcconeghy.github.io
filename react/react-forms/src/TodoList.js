import { useState } from 'react'
import NewTodoForm from "./NewTodoForm";
import Todo from './Todo';
import "./App.css";


function TodoList() {

    const [todoList, setTodoList] = useState([])

    const addTodo = newTodo => {
        setTodoList(todos => [...todos, newTodo])

    }

    const update = (id, updatedText) => {
        setTodoList(todos =>
            todos.map(todo =>
                todo.id === id ? { ...todo, text: updatedText } : todo 
            )
        )
    }

    const remove = id => {
        setTodoList(todos => todos.filter(todo => todo.id !== id))
    }

    const renderTodos = todoList.map(todo => (
        <Todo 
            key={todo.id}
            id={todo.id}
            text={todo.text}
            handleRemove={remove}
            update={update}
        />
    )) 

  return (
  <div>
      <NewTodoForm addTodo = { addTodo }/><br/>
      <div>
          <h4>Your Todos:</h4>
          <div className="TodoList">
            <ul >{renderTodos}</ul>
        </div>
      </div>
  </div>
  )
}

export default TodoList;
