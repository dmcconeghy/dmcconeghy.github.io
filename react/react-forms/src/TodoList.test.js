import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(todoList, text="test tests") {
    const textInput = todoList.getByLabelText("Todo item:");
    fireEvent.change(textInput, {target: { value: text}})
    const submitButton = todoList.getByText("Add it!")
    fireEvent.click(submitButton)
}

it("renders without crashing", function() {
    render(<TodoList />);
  });

it("matches snapshot", function() {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a todo", function() {
    const list = render(<TodoList />)
    addTodo(list)

    //cleared form, added todo
    expect(list.getByLabelText("Todo item:")).toHaveValue("");
    expect(list.getByText("Todo item:")).toBeInTheDocument()
    expect(list.getByText("Edit")).toBeInTheDocument()
    expect(list.getByText("X")).toBeInTheDocument()
})

it("can edit a todo", function() {
    const list = render(<TodoList />);
    addTodo(list);
  
    fireEvent.click(list.getByText("Edit"));
    const editInput = list.getByDisplayValue("test tests");
    fireEvent.change(editInput, { target: { value: "sleep" }});
    fireEvent.click(list.getByText("Update"));
  
    // expect only edited todo to appear
    expect(list.getByText("sleep")).toBeInTheDocument();
    expect(list.queryByText("test tests")).not.toBeInTheDocument();
  });

  it("can delete a todo", function() {
    const list = render(<TodoList />);
    addTodo(list);
  
    fireEvent.click(list.getByText("X"));
  
    // expect todo to be gone
    expect(list.queryByText("test tests")).not.toBeInTheDocument();
  });