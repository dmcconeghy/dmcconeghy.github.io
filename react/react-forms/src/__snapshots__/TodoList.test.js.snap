// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="TodoForm"
    >
      <form>
        <label
          for="todo"
        >
          Todo item:
        </label>
        <input
          id="todo"
          name="todo"
          placeholder="Enter a todo item"
          type="textarea"
          value=""
        />
        <button
          name="submit"
        >
          Add it!
        </button>
      </form>
    </div>
    <br />
    <div>
      <h4>
        Your Todos:
      </h4>
      <div
        class="TodoList"
      >
        <ul />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
